src/odoc/resolver.ml

```diff
@@ -0,0 +1,227 @@
+let load_units_from_files paths =
+  let safe_read file acc =
+    match Compilation_unit.load file with
+    | Ok u -> u :: acc
+    | Error (`Msg msg) ->
+        let warning =
+          Odoc_model.Error.filename_only "%s" msg (Fs.File.to_string file)
+        in
+        prerr_endline (Odoc_model.Error.to_string warning);
+        acc
+  in
+  List.fold_right safe_read paths []
+
+let unit_cache = Hashtbl.create 42
+
+(** Load every units matching a given name. Cached. *)
+let load_units_from_name =
+  let do_load ap target_name =
+    let paths = Accessible_paths.find ap target_name in
+    load_units_from_files paths
+  in
+  fun ap target_name ->
+    try Hashtbl.find unit_cache target_name
+    with Not_found ->
+      let units = do_load ap target_name in
+      Hashtbl.add unit_cache target_name units;
+      units
+
+let rec find_map f = function
+  | [] -> None
+  | hd :: tl -> (
+      match f hd with Some x -> Some (x, tl) | None -> find_map f tl)
+
+let lookup_module_with_digest ap target_name digest =
+  let module_that_match_digest u =
+    match u with
+    | Compilation_unit.Module_content m
+      when Digest.compare m.Odoc_model.Lang.Compilation_unit.digest digest = 0
+      ->
+        Some m
+    | _ -> None
+  in
+  let units = load_units_from_name ap target_name in
+  match find_map module_that_match_digest units with
+  | Some (m, _) -> Odoc_xref2.Env.Found m
+  | None -> Not_found
+
+(** Lookup a module matching a name. If there is more than one result, report on
+    stderr and return the first one.
+
+    TODO: Correctly propagate warnings instead of printing. *)
+let lookup_module_by_name ap target_name =
+  let first_module u =
+    match u with
+    | Compilation_unit.Module_content m -> Some m
+    | Page_content _ -> None
+  in
+  let rec find_ambiguous tl =
+    match find_map first_module tl with
+    | Some (m, tl) -> m :: find_ambiguous tl
+    | None -> []
+  in
+  let units = load_units_from_name ap target_name in
+  match find_map first_module units with
+  | Some (m, tl) ->
+      (match find_ambiguous tl with
+      | [] -> ()
+      | ambiguous ->
+          let ambiguous = m :: ambiguous in
+          let ambiguous =
+            List.map
+              (fun m -> root_name m.Odoc_model.Lang.Compilation_unit.root)
+              ambiguous
+          in
+          let warning =
+            Odoc_model.Error.filename_only
+              "Ambiguous lookup. Possible files: %a"
+              Format.(pp_print_list pp_print_string)
+              ambiguous target_name
+          in
+          prerr_endline (Odoc_model.Error.to_string warning));
+      Odoc_xref2.Env.Found m
+  | None -> Not_found
+
+(** Lookup a module. First looks into [imports_map] then searches into the paths. *)
+let lookup_unit ~important_digests ~imports_map ap target_name =
+  match StringMap.find target_name imports_map with
+  | Odoc_model.Lang.Compilation_unit.Import.Unresolved (_, Some digest) ->
```

Comment:
Are the imports every anything other than `Unresolved` after this patch? If not, can we get rid of `Compilation_unit.Import.t`?

Reply:
Looking at the `loader` code, imports are indeed always `Unresolved`. I'll simplify the type.

Reply:
I'm not convinced `link-deps` can _ever_ produce the correct set of dependencies. In any case, I'm also not convinced the logic you're talking about is sound. It's currently doing this:

```
List.iter odoctree.Odoc_model.Lang.Compilation_unit.imports
        ~f:(fun import ->
          match import with
          | Odoc_model.Lang.Compilation_unit.Import.Unresolved _ -> ()
          | Odoc_model.Lang.Compilation_unit.Import.Resolved (root, _) ->
              Hash_set.add deps root);
```

and iterating this function over all odoc files in a directory. The only thing this gains you over using simply unioning the output of `compile-deps` is that it won't mention modules that couldn't be resolved at `compile` time - ie, modules for which we're missing the `cm{t,ti,i}` file. There's probably one minor difference - if we're making a module that's entirely aliases to other modules, and nothing else within the same directory dereferences those aliases, then we might resolve that at compile time and thus miss it if we don't make a note of resolved imports. This is pretty rare though, and I only know of one instance of this (ocaml-compiler-libs).

The second point is that this entirely ignores all the references in the odoc files, so if there are any references to modules that aren't a compilation dependency then this logic will miss those dependencies. Even if we were to resolve the first part of the reference at compile time (as was suggested in a recent meeting), we'd need to look recursively at the referenced module - e.g. if we had the reference `Stdlib.List.t` and we resolved the `Stdlib` part, we'd need to have a link dependency not just on `stdlib.odoc` but also `stdlib__list.odoc`.

Reply:
Indeed. Ideally, the link-dependencies of a unit is everything else. The smallest reasonable set would be the transitive compile-deps, which the `link-deps` command doesn't even compute.
Do you suggest removing this command ?

> it won't mention modules that couldn't be resolved at compile time

With my proposal it will also remove modules that weren't used at compile time, which may not be what we want.

Reply:
I think that's likely worse, as it's much more likely we'll hit the problem of references above. At least with the 'union of compile-deps' scheme we're not producing _fewer_ link deps than before.

Reply:
I've re-implemented the efficient lookup for imports, the performance regression is gone. The `link-deps` command is still here and has a testcase.